### PR TITLE
Optimize `TyKind::eq`.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

File: compiler/rustc_type_ir/src/sty.rs [hunk 1]
```diff
@@ -310,47 +310,52 @@ impl<I: Interner> Clone for TyKind<I> {
...
+            (Param(a_p), Param(b_p)) => a_p == b_p,
+            (Bound(a_d, a_b), Bound(b_d, b_b)) => a_d == b_d && a_b == b_b,
+            (Placeholder(a_p), Placeholder(b_p)) => a_p == b_p,
+            (Infer(a_t), Infer(b_t)) => a_t == b_t,
+            (Error(a_e), Error(b_e)) => a_e == b_e,
+            (Bool, Bool) | (Char, Char) | (Str, Str) | (Never, Never) => true,
+            _ => {
+                debug_assert!(
+                    tykind_discriminant(self) != tykind_discriminant(other),
+                    "This branch must be unreachable, maybe the match is missing an arm? self = self = {self:?}, other = {other:?}"
```
Comment (fbstj) [inline compiler/rustc_type_ir/src/sty.rs:354]:
```suggestion
                    "This branch must be unreachable, maybe the match is missing an arm? self = {self:?}, other = {other:?}"
```

**Comment by nnethercote** [comment]
Small but clear icount wins here.

r? @compiler-errors

**Comment by compiler-errors** [comment]
r=me after making tidy happy

**Comment by nnethercote** [comment]
I fixed the assertion text and the formatting that tidy complained about.

@bors r=compiler-errors

**Comment by nnethercote** [comment]
Some kind of LLVM build failure, presumably unrelated to the changes in this PR:
```
FAILED: lib/libLLVMMCJIT.a 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E rm -f lib\libLLVMMCJIT.a && D:\a\rust\rust\mingw64\bin\ar.exe qc lib\libLLVMMCJIT.a  lib/ExecutionEngine/MCJIT/CMakeFiles/LLVMMCJIT.dir/MCJIT.cpp.obj && D:\a\rust\rust\mingw64\bin\ranlib.exe lib\libLLVMMCJIT.a && cd ."
D:\a\rust\rust\mingw64\bin\ranlib.exe: could not create temporary file whilst writing archive: no more archived files
```

**Comment by nnethercote** [comment]
On aarch64-gnu it failed, ending with this:
```
== clock drift check ==
  local time: Thu Feb  9 00:02:07 UTC 2023
Session terminated, killing shell... ...killed.
time="2023-02-09T00:02:08Z" level=error msg="error waiting for container: unexpected EOF"
  network time: 
Error: The operation was canceled.
```
Again, this seems entirely unrelated to the changes in the PR. Did I just get unlucky again?

@bors retry

**Comment by compiler-errors** [comment]
Yeah, bors has been pretty flaky recently imo

**Comment by nnethercote** [comment]
keccak and cranelift-codegen are noisy. wg-grammar saw the expected benefit, but it's now considered non-significant, I guess it must have been a bit noisy recently as well.

@rustbot label: +perf-regression-triaged

**Pull request merged**